### PR TITLE
fix(Bugs DT App):fix multi bugs in gui

### PR DESCRIPTION
**Pull request by bahaaabed**
**install The Desktop App:**

- cd biometric-attendance-sync-tool
  && python3 -m venv venv
  && source venv/bin/activate
  && pip install -r requirements.txt

- create 2 folders (local_config.json) and  (local_config.py)

- pip install pyinstaller

- python -m PyInstaller --name="attendance-sync" --windowed --onefile gui.py

- run the exec file in the dist folder

### CODE REVIEW

File: gui.py [hunk 1]
```diff
@@ -342,15 +416,17 @@ def validate_date(date):
...
         return False
 
 
 def convert_into_date(datestring, pattern):
     try:
         return datetime.datetime.strptime(datestring, pattern)
-    except:
+    except Exception as e:
+        print(e)
         return None
```
Comment (hamzaabadah) [inline gui.py:394]: log the Exception

File: erpnext_sync.py [hunk 2]
```diff
@@ -85,7 +94,7 @@ def pull_process_and_push_data(device, device_attendance_logs=None):
     attendance_success_logger = setup_logger(attendance_success_log_file, '/'.join([config.LOGS_DIRECTORY, attendance_success_log_file])+'.log')
     attendance_failed_logger = setup_logger(attendance_failed_log_file, '/'.join([config.LOGS_DIRECTORY, attendance_failed_log_file])+'.log')
     if not device_attendance_logs:
-        device_attendance_logs = get_all_attendance_from_device(device['ip'], device_id=device['device_id'], clear_from_device_on_fetch=device['clear_from_device_on_fetch'])
+        device_attendance_logs = get_all_attendance_from_device(device['ip'], device_id=device['device_id'],password= int(config.DEVICE_PASSWORD), clear_from_device_on_fetch=device['clear_from_device_on_fetch'])
```
Comment (newhr1) [inline erpnext_sync.py]: you need to have device password on the lever of device not direct to the config

File: erpnext_sync.py [hunk 3]
```diff
@@ -124,12 +133,15 @@ def pull_process_and_push_data(device, device_attendance_logs=None):
                 punch_direction = 'IN'
             else:
                 punch_direction = None
-        erpnext_status_code, erpnext_message = send_to_erpnext(device_attendance_log['user_id'], device_attendance_log['timestamp'], device['device_id'], punch_direction)
+        erpnext_status_code, erpnext_message = send_to_erpnext(device_attendance_log['user_id'], device_attendance_log['timestamp'],device_attendance_log['punch'])
```
Comment (newhr1) [inline erpnext_sync.py]: also send the device id here

File: erpnext_sync.py [hunk 4]
```diff
@@ -160,11 +174,12 @@ def get_all_attendance_from_device(ip, port=4370, timeout=30, device_id=None, cl
...
                 f.write(json.dumps(list(map(lambda x: x.__dict__, attendances)), default=datetime.datetime.timestamp))
             if clear_from_device_on_fetch:
-                x = conn.clear_attendance()
+                # x = conn.clear_attendance()
                 info_logger.info("\t".join((ip, "Attendance Clear Attempted. Result:", str(x))))
         x = conn.enable_device()
         info_logger.info("\t".join((ip, "Device Enable Attempted. Result:", str(x))))
-    except:
+    except Exception as e:
+        print(e)
```
Comment (newhr1) [inline erpnext_sync.py]: use the logger please

File: erpnext_sync.py [hunk 5]
```diff
@@ -173,34 +188,41 @@ def get_all_attendance_from_device(ip, port=4370, timeout=30, device_id=None, cl
...
+        
+        else :
+            error_str = _safe_get_error_str(response)
+            if EMPLOYEE_NOT_FOUND_ERROR_MESSAGE in error_str:
+                error_logger.error('\t'.join(['Error during ERPNext API Call.', str(employee_field_value), str(timestamp.timestamp()), str(log_type), error_str]))
+                # TODO: send email?
+            else:
+                error_logger.error('\t'.join(['Error during ERPNext API Call.', str(employee_field_value), str(timestamp.timestamp()), str(log_type), error_str]))
+            return response.status_code, error_str
+    except Exception as e:
```
Comment (newhr1) [inline erpnext_sync.py:213]: use logger please

File: erpnext_sync.py [hunk 6]
```diff
@@ -228,7 +250,8 @@ def update_shift_last_sync_timestamp(shift_type_device_mapping):
                         response_code = send_shift_sync_to_erpnext(shift, min_pull_timestamp)
                         if response_code == 200:
                             status.set(f'{shift}_sync_timestamp', str(min_pull_timestamp))
-                except:
+                except Exception as e:
+                    print(e)
```
Comment (newhr1) [inline erpnext_sync.py]: remove print

File: gui.py [hunk 7]
```diff
@@ -1,16 +1,22 @@
...
 from PyQt5.QtCore import QRegExp
 from PyQt5.QtGui import QIntValidator, QRegExpValidator
 from PyQt5.QtWidgets import QApplication, QLabel, QLineEdit, QMainWindow, QMessageBox, QPushButton
+from functools import partial
+from multiprocessing.pool import Pool
+from queue import Queue
+
 
 
 config_template = '''# ERPNext related configs
```
Comment (newhr1) [inline gui.py:21]: no need for this anyomre

File: gui.py [hunk 8]
```diff
@@ -81,6 +88,10 @@ def setup_textboxes_and_label(self):
                           "import_start_date", 250, 60, 200, 30)
         self.create_field("textbox_import_start_date", 250, 90, 200, 30)
         self.validate_data(r"^\d{1,2}/\d{1,2}/\d{4}$", "textbox_import_start_date")
+
```
Comment (newhr1) [inline gui.py:87]: the same comment before

File: gui.py [hunk 9]
```diff
@@ -236,9 +264,21 @@ def setup_local_config(self):
 
         if os.path.exists("local_config.py"):
             os.remove("local_config.py")
+
```
Comment (newhr1) [inline gui.py]: this method also suppose to be deleted

File: gui.py [hunk 10]
```diff
@@ -342,15 +416,17 @@ def validate_date(date):
     try:
         datetime.datetime.strptime(date, '%d/%m/%Y')
         return True
-    except ValueError:
+    except ValueError as e:
+        print(e)
```
Comment (newhr1) [inline gui.py]: use the logger please

Comment (hamzaabadah) [reply gui.py]: The Gui file does not have a logger

**Pull request merged**